File: android/app/src/main/java/com/borsystems/app/auth/Permissions.kt
package com.borsystems.app.auth

import com.borsystems.app.network.CurrentUser
import com.borsystems.app.network.UserRole

/**
 * Capability layer — Android mirror of web `lib/permissions.tsx`.
 *
 * The signed-in user's *effective* permissions come from GET /users/me
 * (`permissions: Map<String, Boolean>` — module.* + action.* keys; admin has
 * every key true). [Capabilities.of] surfaces them through a `can(key)` helper
 * the nav, the More menu and in-screen actions all consume, so visibility logic
 * lives in one place.
 *
 * If an older backend session hasn't sent a permission map, we derive a safe
 * baseline from the user's role (the same DEFAULT_PERMISSIONS the server uses),
 * so nothing is accidentally hidden for admins.
 */

/** Module-visibility keys, mirrored from the backend catalogue. */
object Module {
    const val OPERATIONS = "module.operations"
    const val MAINTENANCE = "module.maintenance"
    const val COMPLIANCE = "module.compliance"
    const val BUSINESS = "module.business"
    const val INSIGHTS = "module.insights"
    const val ADMIN = "module.admin"
}

/** Sensitive-action keys, mirrored from the backend catalogue. */
object Action {
    const val APPROVE_PERMITS = "action.approve_permits"
    const val APPROVE_QUOTES = "action.approve_quotes"
    const val EDIT_COMPLIANCE = "action.edit_compliance"
    const val MANAGE_DEVICES = "action.manage_devices"
    const val MANAGE_AUTOMATIONS = "action.manage_automations"
    const val EXPORT_REPORTS = "action.export_reports"
    const val MANAGE_USERS = "action.manage_users"
    const val MANAGE_BILLING = "action.manage_billing"
    const val DELETE_RECORDS = "action.delete_records"
}

private val ALL_MODULES = listOf(
    Module.OPERATIONS, Module.MAINTENANCE, Module.COMPLIANCE,
    Module.BUSINESS, Module.INSIGHTS, Module.ADMIN,
)
private val ALL_ACTIONS = listOf(
    Action.APPROVE_PERMITS, Action.APPROVE_QUOTES, Action.EDIT_COMPLIANCE,
    Action.MANAGE_DEVICES, Action.MANAGE_AUTOMATIONS, Action.EXPORT_REPORTS,
    Action.MANAGE_USERS, Action.MANAGE_BILLING, Action.DELETE_RECORDS,
)
private val ALL_KEYS = ALL_MODULES + ALL_ACTIONS

private fun mapOfAll(value: Boolean): Map<String, Boolean> =
    ALL_KEYS.associateWith { value }

/**
 * Per-role default baselines — a client mirror of DEFAULT_PERMISSIONS on the
 * server. Only used to approximate what a role can see when the server hasn't
 * supplied a real map. The authoritative maps still come from /users/me.
 */
val DEFAULT_PERMISSIONS: Map<UserRole, Map<String, Boolean>> = mapOf(
    UserRole.admin to mapOfAll(true),
    UserRole.supervisor to (mapOfAll(false) + mapOf(
        Module.OPERATIONS to true,
        Module.MAINTENANCE to true,
        Module.COMPLIANCE to true,
        Module.BUSINESS to true,
        Module.INSIGHTS to true,
        Module.ADMIN to false,
        Action.APPROVE_PERMITS to true,
        Action.APPROVE_QUOTES to true,
        Action.EDIT_COMPLIANCE to true,
        Action.MANAGE_DEVICES to true,
        Action.MANAGE_AUTOMATIONS to false,
        Action.EXPORT_REPORTS to true,
        Action.MANAGE_USERS to false,
        Action.MANAGE_BILLING to false,
        Action.DELETE_RECORDS to true,
    )),
    UserRole.cleaner to (mapOfAll(false) + mapOf(
        Module.OPERATIONS to true,
    )),
)

/**
 * Resolved capabilities for one user. Build with [of]; query with [can].
 *
 * `can` honours the backend map when present. Admin is always allowed
 * (defensive — the server agrees), even if a key is somehow missing.
 */
class Capabilities private constructor(
    val role: UserRole,
    private val effective: Map<String, Boolean>,
) {
    fun can(key: String): Boolean {
        if (role == UserRole.admin) return true
        return effective[key] == true
    }

    /** True for admin or supervisor — the "staff" tier (vs field cleaners). */
    val isStaff: Boolean get() = role == UserRole.admin || role == UserRole.supervisor
    val isAdmin: Boolean get() = role == UserRole.admin

    companion object {
        /** Empty capabilities for a logged-out state — everything denied. */
        val NONE = Capabilities(UserRole.cleaner, emptyMap())

        fun of(user: CurrentUser?): Capabilities {
            if (user == null) return NONE
            val effective = if (user.permissions.isNotEmpty()) {
                user.permissions
            } else {
                // Older session without a map — fall back to the role baseline.
                DEFAULT_PERMISSIONS[user.role] ?: emptyMap()
            }
            return Capabilities(user.role, effective)
        }
    }
}

File: ios/BORSystems/Permissions.swift
import Foundation

/// Honour the backend's permission model on the phone.
///
/// The web app gates nav sections and sensitive actions on a `permissions`
/// map the backend computes per (org, role): `module.*` keys decide which
/// sections of the app are visible, `action.*` keys decide which sensitive
/// operations a user may perform (see backend/services/permissions.ts).
///
/// `Capabilities` is the single source of truth the iOS UI asks before showing
/// a tab, a More-hub row, or an action button. It PREFERS the live backend map
/// when `GET /users/me` supplied one, and otherwise falls back to a role→
/// capability map that mirrors the server's `DEFAULT_PERMISSIONS` baseline so
/// the app still gates correctly against the leaner login / 2FA payloads (which
/// omit `permissions`).

// MARK: - Permission keys

/// The recognised permission keys, mirroring `PERMISSION_KEYS` on the backend.
/// Using an enum (rather than bare strings at call sites) keeps the gates
/// typo-proof and greppable.
enum Permission: String, CaseIterable {
    // module.* — which sections of the app nav a role can see.
    case moduleOperations  = "module.operations"
    case moduleMaintenance = "module.maintenance"
    case moduleCompliance  = "module.compliance"
    case moduleBusiness    = "module.business"
    case moduleInsights    = "module.insights"
    case moduleAdmin       = "module.admin"

    // action.* — sensitive actions a role may perform.
    case approvePermits    = "action.approve_permits"
    case approveQuotes     = "action.approve_quotes"
    case editCompliance    = "action.edit_compliance"
    case manageDevices     = "action.manage_devices"
    case manageAutomations = "action.manage_automations"
    case exportReports     = "action.export_reports"
    case manageUsers       = "action.manage_users"
    case manageBilling     = "action.manage_billing"
    case deleteRecords     = "action.delete_records"
}

// MARK: - Role baseline (fallback)

/// Per-role baselines, mirroring `DEFAULT_PERMISSIONS` on the backend. Used only
/// when the live `permissions` map is absent from the current user (older or
/// leaner payloads). When the backend map is present it always wins.
private let defaultPermissionsByRole: [UserRole: Set<Permission>] = [
    // admin — everything (also short-circuited as always-allowed below).
    .admin: Set(Permission.allCases),
    // supervisor — every module except admin; operational actions + exports,
    // but not the org-governance actions (users / billing / automations).
    .supervisor: [
        .moduleOperations, .moduleMaintenance, .moduleCompliance,
        .moduleBusiness, .moduleInsights,
        .approvePermits, .approveQuotes, .editCompliance,
        .manageDevices, .exportReports, .deleteRecords,
    ],
    // cleaner — "field staff": the operations section only, no sensitive actions.
    .cleaner: [
        .moduleOperations,
    ],
]

// MARK: - Capabilities

/// What the signed-in user is allowed to see and do. Build it from the current
/// user, then ask `can(_:)` at every gate.
struct Capabilities {
    let role: UserRole
    /// The live backend map keyed by raw permission string, if one was supplied.
    private let backend: [String: Bool]?

    init(user: CurrentUser?) {
        self.role = user?.role ?? .cleaner
        self.backend = user?.permissions
    }

    /// Whether the user holds a given permission. Admin is always allowed
    /// (matching the server, which never gates admin). Otherwise prefer the
    /// backend map; fall back to the role baseline when it's absent.
    func can(_ permission: Permission) -> Bool {
        if role == .admin { return true }
        if let backend, let value = backend[permission.rawValue] {
            return value
        }
        return defaultPermissionsByRole[role]?.contains(permission) ?? false
    }

    // MARK: Convenience gates used by the nav + screens

    /// Cleaning operations: live alerts, floor plans, dispatch, shifts.
    var canSeeOperations: Bool { can(.moduleOperations) }
    /// Work orders, meters, PPMs, contractors.
    var canSeeMaintenance: Bool { can(.moduleMaintenance) }
    /// SDS, certifications, inspections.
    var canSeeCompliance: Bool { can(.moduleCompliance) }
    /// Analytics, reports, dashboards.
    var canSeeInsights: Bool { can(.moduleInsights) }
    /// Settings, users, billing, devices, automations.
    var canSeeAdmin: Bool { can(.moduleAdmin) }

    /// Register / decommission hangers + gateways.
    var canManageDevices: Bool { can(.manageDevices) }
    /// Create / deactivate / delete staff.
    var canManageUsers: Bool { can(.manageUsers) }
    /// Download CSV / PDF exports.
    var canExportReports: Bool { can(.exportReports) }
    /// Hard-delete records.
    var canDeleteRecords: Bool { can(.deleteRecords) }

    /// Whether the user gets the org-wide dispatch SEND form (admins +
    /// supervisors) versus the read-only "dispatches sent to me" inbox.
    /// Maps to the operations module plus a coordinator role — cleaners
    /// only ever receive dispatches.
    var canSendDispatch: Bool {
        canSeeOperations && (role == .admin || role == .supervisor)
    }

    /// Whether to surface the management / insights / system hub on the More
    /// tab at all. Cleaners get a slimmed hub (profile + sign out only).
    var hasManagementHub: Bool {
        canSeeMaintenance || canSeeInsights || canSeeAdmin || canManageDevices
    }
}

extension AuthStore {
    /// Capabilities for the currently signed-in user. Recomputed each access so
    /// it always reflects the latest `user` snapshot.
    var capabilities: Capabilities { Capabilities(user: user) }
}
